compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt

```diff
@@ -0,0 +1,405 @@
+        val (x, overscrollX) = availableDelta(delta.x, overscrollOffset.x, source)
+        val (y, overscrollY) = availableDelta(delta.y, overscrollOffset.y, source)
+
+        overscrollOffset = Offset(overscrollX, overscrollY)
+
+        return Offset(x, y)
+    }
+
+    /*
+     * Semantics of this method match the [OverscrollEffect.applyToScroll] one,
+     * The only difference is NestedScrollSource being remapped to CupertinoScrollSource to narrow
+     * processed states invariant
+     */
+    private fun applyToScroll(
+        delta: Offset,
+        source: CupertinoScrollSource,
+        performScroll: (Offset) -> Offset
+    ): Offset {
+        // Calculate how much delta is available after being consumed by scrolling inside overscroll area
+        val deltaLeftForPerformScroll = availableDelta(delta, source)
+
+        // Then pass remaining delta to scroll closure
+        val deltaConsumedByPerformScroll = performScroll(deltaLeftForPerformScroll)
+
+        // Delta which is left after `performScroll` was invoked with availableDelta
+        val unconsumedDelta = deltaLeftForPerformScroll - deltaConsumedByPerformScroll
+
+        return when (source) {
+            CupertinoScrollSource.DRAG -> {
+                // [unconsumedDelta] is going into overscroll again in case a user drags and hits the
+                // overscroll->content->overscroll or content->overscroll scenario within single frame
+                overscrollOffset += unconsumedDelta
+
+                lastFlingUncosumedDelta = Offset.Zero
+
+                delta - unconsumedDelta
+            }
+
+            CupertinoScrollSource.FLING -> {
+                // If unconsumedDelta is not Zero, [CupertinoFlingEffect] will cancel fling and
+                // start spring animation instead
+                lastFlingUncosumedDelta = unconsumedDelta
+
+                delta - unconsumedDelta
+            }
+        }
+    }
+
+    override fun applyToScroll(
+        delta: Offset,
+        source: NestedScrollSource,
+        performScroll: (Offset) -> Offset
+    ): Offset {
+        direction = direction.combinedWith(delta.toCupertinoOverscrollDirection())
+
+        return source.toCupertinoScrollSource()?.let {
+            applyToScroll(delta, it, performScroll)
+        } ?: performScroll(delta)
+    }
+
+    override suspend fun applyToFling(
+        velocity: Velocity,
+        performFling: suspend (Velocity) -> Velocity
+    ) {
+        val availableFlingVelocity = playInitialSpringAnimationIfNeeded(velocity)
+
+        val velocityConsumedByFling = performFling(availableFlingVelocity)
+        val postFlingVelocity = availableFlingVelocity - velocityConsumedByFling
+
+        playSpringAnimation(
+            lastFlingUncosumedDelta.toFloat(),
+            postFlingVelocity.toFloat(),
+            false
+        )
+    }
+
+    private fun Offset.toCupertinoOverscrollDirection(): CupertinoOverscrollDirection {
+        val epsilon = 1E-4f
```

Comment:
const?
Is there anything to reuse in math utils?

Reply:
I double-checked how Offset is constructed before passed here, it is generated from 0f literal, so I just replaced those with strict 0f comparison.